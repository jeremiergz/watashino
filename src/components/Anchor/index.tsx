import Box, { BoxProps } from 'components/primitives/Box';
import { Theme } from 'components/providers/ThemeProvider';
import React from 'react';
import styled from 'styled-components';
import { variant } from 'styled-system';

const A = styled(Box)`
  color: inherit;
  text-decoration: none;
  transition: color 100ms ease-in-out, filter 100ms ease-in-out;
  will-change: filter;
  &:hover {
    color: ${({ theme }: { theme: Theme }) => theme.colors.primary};
    filter: brightness(0.75);
  }
  ${variant({
    variants: {
      underlined: {
        textDecoration: 'underline',
      },
    },
  })}
`;

const Anchor: React.FC<AnchorProps> = ({ children, ...rest }) => {
  const isTouchDevice = typeof window !== 'undefined' && window.navigator.maxTouchPoints > 0;
  const target = isTouchDevice ? '_self' : '_blank';
  return (
    <A as="a" fontWeight="semiBold" target={target} {...rest}>
      {children}
    </A>
  );
};

Anchor.displayName = 'Anchor';

export type AnchorProps = BoxProps & {
  href?: string;
  target?: string;
  variant?: 'underlined';
};
export default Anchor;
