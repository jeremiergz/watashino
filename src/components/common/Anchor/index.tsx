import React from 'react';
import styled from 'styled-components';
import { variant } from 'styled-system';
import BaseBox, { BoxProps } from '../../primitives/Box';

type AnchorProps = BoxProps & {
  download?: boolean;
  href?: string;
  target?: string;
  variant?: 'underlined';
};

const A = styled(BaseBox)`
  color: inherit;
  text-decoration: none;
  transition: color 100ms ease-in-out;
  .gatsby-image-wrapper {
    transition: filter 100ms ease-in-out;
  }
  &:hover {
    color: ${({ theme }) => theme.colors.primary};
    .gatsby-image-wrapper {
      filter: brightness(0.75);
    }
  }
  ${variant({
    variants: {
      underlined: {
        textDecoration: 'underline',
      },
    },
  })}
`;

const Anchor = ({ children, ...rest }: AnchorProps) => {
  const isTouchDevice = typeof window !== 'undefined' && window.navigator.maxTouchPoints > 0;
  const target = isTouchDevice ? '_self' : '_blank';
  return (
    <A as="a" fontWeight="semi-bold" target={target} {...rest}>
      {children}
    </A>
  );
};

Anchor.displayName = 'Anchor';

export { AnchorProps };
export default Anchor;
