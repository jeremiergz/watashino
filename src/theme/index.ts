const breakpoints = ['320px', '375px', '425px', '768px', '1024px', '1200px', '1440px', '2560px'] as const;
Object.defineProperties(breakpoints, {
  mobileS: { value: breakpoints[0] },
  mobileM: { value: breakpoints[1] },
  mobileL: { value: breakpoints[2] },
  tablet: { value: breakpoints[3] },
  laptopS: { value: breakpoints[4] },
  laptopM: { value: breakpoints[5] },
  laptopL: { value: breakpoints[6] },
  desktop: { value: breakpoints[7] },
});

const colors = {
  black: '#212121',
  dark: '#424242',
  error: '#F44336',
  gray: '#757575',
  info: '#2196F3',
  lightgray: '#E0E0E0',
  primary: '#31859A',
  secondary: '#17365C',
  success: '#4CAF50',
  white: '#FAFAFA',
} as const;

const fonts = {
  main: "'Open Sans', sans-serif",
} as const;

const fontWeights = {
  bold: 800,
  regular: 400,
  'semi-bold': 600,
} as const;

const DarkTheme = {
  background: colors.black,
  breakpoints,
  colors: {
    ...colors,
    text: colors.white,
  },
  fonts,
  fontWeights,
  svg: colors.white,
  type: 'dark',
};

const LightTheme = {
  background: colors.white,
  breakpoints,
  colors: {
    ...colors,
    text: '#262626',
  },
  fonts,
  fontWeights,
  svg: '#262626',
  type: 'light',
};

type Theme = typeof DarkTheme | typeof LightTheme;
type ThemeProps = {
  background?: string;
  colors?: keyof Theme['colors'] | 'inherit';
  fontWeight?: keyof Theme['fontWeights'];
  svg?: string;
  type?: 'light' | 'dark';
};

export { colors, DarkTheme, LightTheme, Theme, ThemeProps };
