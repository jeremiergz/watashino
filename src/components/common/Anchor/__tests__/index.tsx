import React from 'react';
import ReactTestRenderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';
import Anchor from '..';
import { DarkTheme } from '../../../../theme';

describe('components/common/Link Test Suite', () => {
  it('matches snapshot', () => {
    const jsx = (
      <ThemeProvider theme={DarkTheme}>
        <Anchor variant="underlined">Anchor Component</Anchor>
      </ThemeProvider>
    );
    expect(ReactTestRenderer.create(jsx).toJSON()).toMatchSnapshot();
  });

  it('should use target="_blank" on desktop devices by default', () => {
    Object.assign(window.navigator, { maxTouchPoints: 0 });
    const jsx = (
      <ThemeProvider theme={DarkTheme}>
        <Anchor variant="underlined">Anchor Component</Anchor>
      </ThemeProvider>
    );
    const renderedInstance = ReactTestRenderer.create(jsx).root;
    expect(renderedInstance.findByType('a').props.target).toBe('_blank');
  });

  it('should use target="_self" on touch devices by default', () => {
    Object.assign(window.navigator, { maxTouchPoints: 4 });
    const jsx = (
      <ThemeProvider theme={DarkTheme}>
        <Anchor variant="underlined">Anchor Component</Anchor>
      </ThemeProvider>
    );
    const renderedInstance = ReactTestRenderer.create(jsx).root;
    expect(renderedInstance.findByType('a').props.target).toBe('_self');
  });
});
