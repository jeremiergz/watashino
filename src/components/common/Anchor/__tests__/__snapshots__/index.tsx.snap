// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/widgets/Link Test Suite matches snapshot 1`] = `
<a
  className="sc-bdVaJa sc-bwzfXH kSBVxK"
  href="/test-page"
  target="_blank"
  variant="underlined"
>
  Link Component
</a>
`;
